### DISCUSSION TITLE
External IP & detect-external-ip question

### DISCUSSION THREAD

**Discussion by Maypul** [coturn/coturn#1112] [Q&A]
Hi there, I am trying to understand how coturn works. From what I read, it needs external/public IP (the one that anyone from the internet could connect to me).
But then I try to use detect-external-ip in docker and it seems to always returning local IP addresses. Now I am not sure if Iunderstand it correctly or maybe it does not work. If it does not work ,how could I "improve" it so it detects my public IP?
If I use with ports mapped in docker, it detects docker container IP. If I map container as host, it detects host local ip in addition.
My setup is:
Internet <> router <> router <> host <> docker
Also I am behind cloudflare but I am guessing I would need to use my IP, not proxied from CF? But then, how would I achieve to have it automatic each time container starts (and possibly reboots on change)?

**Comment by tyranron** [comment]
@Maypul there is no mystery in how `detect-external-ip` works, its [source is open and available for reading](https://github.com/coturn/coturn/blob/docker/4.6.0-r1/docker/coturn/rootfs/usr/local/bin/detect-external-ip.sh).

You may always debug how it works via simple command:
```shell
docker run --rm coturn/coturn:alpine detect-external-ip
```

As you're behind CloudFlare, I guess that the DNS way, used by it, just doesn't work for you:
> ```bash
> #/ Typically it is possible to make an HTTP request to a number of providers
> #/ that offer the external IP in their response body (eg: ifconfig.me). However,
> #/ why do a slow and heavy HTTP request, when DNS exists and is much faster?
> #/ Well established providers such as OpenDNS or Google offer special hostnames
> #/ that, when resolved, will actually return the IP address of the caller.
> ```

You may always write your own `detect-external-ip.sh` script, with the logic you need, and mount it like this (ensure it has execution rights before mounting):
```shell
docker run --rm -v /path/to/my-detect-external-ip.sh:/usr/local/bin/detect-external-ip.sh coturn/coturn:alpine
```

However, I guess, the alternative HTTP way (like via `ifconfig.me`) won't work either, because CloudFlare is not a NAT for you, but rather is an HTTP proxy server.

Moreover, TURN/STUN protocols are not HTTP, so [CloudFlare is not able to proxy them](https://community.cloudflare.com/t/stun-udp-3478-to-unifi-controller-blocked-by-cloudflare/65084/6). So you simply cannot hide Coturn behind CloudFlare, despite the solution of the original question.
